### ISSUE TITLE
Onboarding: interactive addon tours

### ISSUE THREAD

**Issue by lisa-wolfgang** [ScratchAddons/ScratchAddons#5407]
Original idea from W_L: https://discord.com/channels/751206349614088204/751417008113057812/1048070490192871455

Basically, the idea is to allow the user to temporarily "enable" addons in the editor or on the website. This would allow us to expose our users to more addons in a more natural environment.

Possible ways this is initiated:
- As a part of onboarding, where we automatically open a project specifically designed to show off our addons (potentially including enabled by default addons, so the user could choose to disable them -- this would allow us to enable more addons by default)
- As a part of normal Scratch usage, where we would occasionally show popups when the user performs an action related to a recommended or featured addon (key word is "occasionally" to prevent annoying the user -- would need to be ratelimited)
- As a "tour" of the new features whenever an update is released

No matter how this is implemented, it needs to include an option to skip as well as disable the tours. We also want to avoid being too annoying -- the user should always have an easy way out of this feature. Perhaps after enabling an addon, it asks if you'd like to try another one or stop for now.
Also, the Scratch editor might take a long time to load for some users, so they might exit the onboarding because of that. Maybe we could begin loading the editor in the background.

**Comment by BroJac5246** [comment]
> As a part of onboarding, where we automatically open a project specifically designed to show off our addons (potentially including enabled by default addons, so the user could choose to disable them -- this would allow us to enable more addons by default)

Or possibly a "Check it out!" button by each addon, which will result in us opening up an example project and enabling the addon. When the example project is closed or the onboarding site is refocused, the addon will be disabled (unless it's on by default), and, back at the onboarding site, they'll see a popup asking if they'd like to have the addon enabled.
> As a part of normal Scratch usage, where we would occasionally show popups when the user performs an action related to a recommended or featured addon (key word is "occasionally" to prevent annoying the user -- would need to be ratelimited)

Absolutely not, that's way too intrusive and feels like the sort of thing that a pushy, well-known, monetizing browser extension would do!
> As a "tour" of the new features whenever an update is released

This would be great as part of the "Scratch Addons has been updated!" popup.

**Comment by DNin01** [comment]
> * As a part of onboarding, where we automatically open a project specifically designed to show off our addons (potentially including enabled by default addons, so the user could choose to disable them -- this would allow us to enable more addons by default)

This could be interesting but might be difficult with Scratch's browser extension policy. Maybe it would open a project _not designed_ to show off addons but is a good example of the addons being demonstrated. We could then show floating popups around different areas to show what the addons are doing (e.g. describing a button added by an addon)

> * As a part of normal Scratch usage, where we would occasionally show popups when the user performs an action related to a recommended or featured addon (key word is "occasionally" to prevent annoying the user -- would need to be ratelimited)

This sounds kind of interesting. Wouldn't it be cool if you add line breaks to one of your comments, then SA came up and asked you if you want to enable the _Line breaks in comments_ addon? Anything like this can still be annoying, though - as BroJac5246 says, it feels like the kind of thing bigger companies would do in their browser extensions. SA might feel completely different if it sometimes came up and invited me to an addon all by itself.

Maybe SA could monitor some user activity and offer relevant suggestions on the settings page instead? Like, for instance, if your projects have a lot of blocks in them, it might offer a suggestion for that block searching addon. I'd be fine with that. We might have to change the privacy policy if we do that, though, and it may impact performance.

> * As a "tour" of the new features whenever an update is released

Yeah! Definitely run down the new features.

> Also, the Scratch editor might take a long time to load for some users, so they might exit the onboarding because of that. Maybe we could begin loading the editor in the background.

SA could just load a small project. I'd also like if it loaded a particular project from somebody else, not a new project or one of yours.

---

I also think a "Try out this addon" button would be a nice way to tutorialize addons, like BroJac5246 said.

**Comment by lisa-wolfgang** [comment]
We could bundle the example project with the extension like we do for the default project addon.

I'd like to reiterate that this feature is intended to be as minimally pushy/annoying/"corporate" as possible. Any popup will always include a "don't show suggestions" option or something similar. I could also imagine such behavior being opt-in.

**Comment by BroJac5246** [comment]
> This could be interesting but might be difficult with Scratch's browser extension policy. Maybe it would open a project not designed to show off addons but is a good example of the addons being demonstrated.

Or we could have a dedicated project for Scratch Addons that just doesn't say "Scratch Addons" on it, to still comply with the policy. It could just be called "Example Project", and it would fit the context while not breaking the policy.

> Also, the Scratch editor might take a long time to load for some users, so they might exit the onboarding because of that. Maybe we could begin loading the editor in the background.

Could we store a (small) sample `.sb3` file in the extension's files and then load that instead of waiting for a project's assets to download? Could we combine that with your suggestion?

> Wouldn't it be cool if you add line breaks to one of your comments, then SA came up and asked you if you want to enable the Line breaks in comments addon?

But then you have a script (essentially an additional addon) running in the background, which might affect performance. It's an interesting idea, though, if we get it right.

**Comment by DNin01** [comment]
> > Also, the Scratch editor might take a long time to load for some users, so they might exit the onboarding because of that. Maybe we could begin loading the editor in the background.
> 
> Could we store a (small) sample `.sb3` file in the extension's files and then load that instead of waiting for a project's assets to download? Could we combine that with your suggestion?

This would probably create a new project and save it to your account, which isn't really something I want SA doing.

**Comment by BroJac5246** [comment]
> > > Also, the Scratch editor might take a long time to load for some users, so they might exit the onboarding because of that. Maybe we could begin loading the editor in the background.
> > 
> > Could we store a (small) sample `.sb3` file in the extension's files and then load that instead of waiting for a project's assets to download? Could we combine that with your suggestion?
> 
> This would probably create a new project and save it to your account, which isn't really something I want SA doing.

We could disable autosave.

**Comment by DNin01** [comment]
> We could disable autosave.

Even if we disable autosave, I'm not sure if we could prevent the project from being "created".

**Comment by CST1229** [comment]
> > We could disable autosave.
> 
> Even if we disable autosave, I'm not sure if we could prevent the project from being "created".

We could load some project ID (like 104) but actually load a different project instead of the normal project at that ID, but then the user might end up accidentally changing the project if they happen to own it.

**Comment by BroJac5246** [comment]
> > > We could disable autosave.
> > 
> > Even if we disable autosave, I'm not sure if we could prevent the project from being "created".
> 
> We could load some project ID (like 104) but actually load a different project instead of the normal project at that ID, but then the user might end up accidentally changing the project if they happen to own it.

So if we combine that with no autosave (or no saving in general), if could work.

**Comment by penguinmoose** [comment]
Because opening an example project may be confusing, I think SA has to be completely clear about why it is opening up a project.

**Comment by BroJac5246** [comment]
> Because opening an example project may be confusing, I think SA has to be completely clear about why it is opening up a project.

Could we embed it instead?

**Comment by WorldLanguages** [comment]
> Because opening an example project may be confusing, I think SA has to be completely clear about why it is opening up a project.

It will probably be an embed.

**Comment by penguinmoose** [comment]
> > Because opening an example project may be confusing, I think SA has to be completely clear about why it is opening up a project.
> 
> Could we embed it instead?

Yes, embedding it is a good idea. We'll open up a page, then put an embedded project in it while still showing that it is a SA tour.

**Comment by penguinmoose** [comment]
I just tried turning on `disable-auto-save` and tried creating a new project by going to `scratch.mit.edu/projects/editor`. The project was still saved and created, so we can't just use the disable autosave addon. Instead, we may need to disable all kinds of saving completely (which will not become an addon - simply a change that happens on that specific tour page.

Alternatively, we can somehow sign out the user just for the onboarding tour page. We might be able to do this by [disabling cookies within the embedded editor](https://stackoverflow.com/questions/44837450/recommended-method-to-prevent-any-content-inside-iframe-from-setting-cookies#answer-44837553).

Or, we can simply use an example project we create.

**Comment by hueychen27** [comment]
What about making videos showcasing new features in the addon update popup?
